File: Product_of_Array_Except_Self/solution2.py
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """

    def _solution2(self, nums):
        """
        First loop, for the output[i], calculate the products on the left side.
        Second loop, for the output[i], calculate the products on the right side multiplied the products on the left side
        """
        numsLen = len(nums)
        ans = [1] * numsLen
        leftProducts = 1
        for i in range(1, numsLen):
           leftProducts = ans[i] = leftProducts * nums[i - 1]
            
        rightProducts = 1
        for i in range(numsLen - 2, -1, -1):
            rightProducts = rightProducts * nums[i + 1]
            ans[i] *= rightProducts
        
        return ans
    

File: Wiggle_Sort/solution2.js
function solution(arr) {
  const N = arr.length;
  let left = N % 2 === 0 ? 0 : 1;
  let right = N - 1;
  
  arr.sort((a ,b) => a - b);
  
  while (left < right) {
    const tmp = arr[left];
    arr[left] = arr[right];
    arr[right] = tmp;
    left += 2;
    right -= 2;
  }
}
